### PR TITLE
Intake | Handle submission of Ramp Refiling - PART ONE

### PR DESCRIPTION
**Pull request by shanear**
The submission logic is a doozy, so I split this up. Apologies on this being a little long, I didn't expect it to be so complicated.

## Testing Instructions
- Attempt to finish a RAMP Refiling with out checking anything, you should see error messages on both the confirmation and the issues selection control.
- Attempt to finish a RAMP refiling with selecting the appropriate options. Validate that they are saved in the DB (it also will create an EP, which can be validated in UAT, but it's probably best to wait till we are also making the contentions)

## Notes
All the AC should be satisfied except:
- creating the contentions, which will be handled in a future PR
- If the selection includes at least one eligible contention, navigates to success confirmation page (#4064)
- If the user only selects the ineligible checkbox, navigate to the ineligible confirmation page (#4206)
- If this is an appeal request, navigate user to "appeal created" confirmation page (#4181)

These will be handled in future PRs

connects #3966

### CODE REVIEW

File: app/models/ramp_election.rb [hunk 1]
```diff
@@ -1,27 +1,11 @@
 class RampElection < RampReview
-  class InvalidEndProductError < StandardError; end
-  class EstablishedEndProductNotFound < StandardError; end
-
-  attr_reader :saving_receipt
```
Comment (shanear) [inline app/models/ramp_election.rb]: This was unused. Everything else in this file was moved to `ramp_review.rb` so it could be shared with `ramp_refiling.rb`

File: app/models/ramp_refiling.rb [hunk 2]
```diff
@@ -3,10 +3,34 @@ class RampRefiling < RampReview
 
   validate :validate_receipt_date, :validate_option_selected
 
+  def create_issues!(source_issue_ids:)
+    issues.destroy_all unless issues.empty?
```
Comment (shanear) [inline app/models/ramp_refiling.rb]: this may be unnecessary, there should be no issues on the Ramp Refiling at this point, but I decided to add it in just to be safe.

File: app/models/ramp_refiling.rb [hunk 3]
```diff
@@ -3,10 +3,34 @@ class RampRefiling < RampReview
...
   end
 
+  def needs_end_product?
+    option_selected != "appeal"
+  end
+
+  private
+
+  def create_contentions!
+    # TODO: Create contentions on the end product from the issues
```
Comment (shanear) [inline app/models/ramp_refiling.rb]: Stay tuned for Part 2

File: app/models/ramp_refiling_intake.rb [hunk 4]
```diff
@@ -23,6 +23,17 @@ def review!(request_params)
...
 
+  def complete!(request_params)
+    transaction do
+      detail.create_issues!(source_issue_ids: request_params[:issue_ids])
+      detail.update!(has_ineligible_issue: request_params[:has_ineligible_issue])
+
+      complete_with_status!(:success)
+
+      detail.create_end_product_and_contentions! if detail.needs_end_product?
+    end
```
Comment (shanear) [inline app/models/ramp_refiling_intake.rb]: I know the doing all the VBMS things inside the transaction has a shot at timing it out and blowing everything up. I'm thinking it's safer to do that than risk the data integrity issues with doing it outside of the transaction. We've only seen 2 timeouts in prod, and both would have been problematic either way.

I'm going to keep a close eye on errors coming from this function (as I have for RAMP v1). I'm also thinking about doing some performance tests with this to make sure.

File: app/models/ramp_review.rb [hunk 5]
```diff
@@ -1,4 +1,7 @@
 class RampReview < ActiveRecord::Base
+  class EstablishedEndProductNotFound < StandardError; end
+  class InvalidEndProductError < StandardError; end
```
Comment (shanear) [inline app/models/ramp_review.rb]: everything in here is word for word copied from `ramp_election.rb`

File: client/app/intake/actions/rampRefiling.js [hunk 6]
```diff
@@ -97,3 +97,75 @@ export const setOutsideCaseflowStepsConfirmed = (isConfirmed) => ({
...
+  if (hasError) {
+    return Promise.resolve(false);
+  }
+
+  dispatch({
+    type: ACTIONS.COMPLETE_INTAKE_START,
+    meta: { analytics }
+  });
+
+  const data = {
```
Comment (shanear) [inline client/app/intake/actions/rampRefiling.js]: decided to compile the data for the request inline here, but if there's a better place, lmk

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 7]
```diff
@@ -2,24 +2,36 @@ import React from 'react';
...
+import { setIssueSelected, setHasIneligibleIssue, setOutsideCaseflowStepsConfirmed, completeIntake,
+  processFinishError } from '../../actions/rampRefiling';
 import _ from 'lodash';
 
 class Finish extends React.PureComponent {
   onCheckIssue = (issueId) => (checked) => this.props.setIssueSelected(issueId, checked)
 
+  componentDidUpdate () {
+    if (this.props.outsideCaseflowStepsError && !this.props.finishErrorProcessed) {
+      this.outsideCaseflowStepsNode.scrollIntoView();
```
Comment (shanear) [inline client/app/intake/pages/rampRefiling/finish.jsx]: this is a similar pattern to certification, and seems to work well

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 8]
```diff
@@ -74,7 +86,7 @@ class Finish extends React.PureComponent {
     return <div>
       <h1>Finish processing RAMP Selection form</h1>
 
-      <ol className="cf-bare-list">
+      <ol className="cf-bare-list" ref={(node) => this.outsideCaseflowStepsNode = node}>
```
Comment (shanear) [inline client/app/intake/pages/rampRefiling/finish.jsx]: suggested from Dan Abramov himself
https://github.com/yannickcr/eslint-plugin-react/issues/678#issue-165177220

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 9]
```diff
@@ -119,26 +132,32 @@ class Finish extends React.PureComponent {
...
-      />
-
-      <Table
-        className="cf-table-selections"
-        columns={otherQuestionColumns}
-        rowObjects={otherQuestions}
-        summary={otherQuestionColumns[0].header}
-        getKeyForRow={(_index, question) => question.id}
-      />
+      <div className={issuesSelectedError ? 'usa-input-error' : ''}>
```
Comment (shanear) [inline client/app/intake/pages/rampRefiling/finish.jsx]: looks like this:
screen shot 2018-01-03 at 7 11 10 pm

File: db/migrate/20180103205014_remove_null_constraint_on_contention_reference_id.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class RemoveNullConstraintOnContentionReferenceId < ActiveRecord::Migration
+  def change
+    change_column_null :ramp_issues, :contention_reference_id, true
```
Comment (shanear) [inline db/migrate/20180103205014_remove_null_constraint_on_contention_reference_id.rb]: these new issues won't have contentions initially, so allow this to be nil

File: db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class AddSourceIssueIdToRampIssues < ActiveRecord::Migration
+  def change
+    add_column :ramp_issues, :source_issue_id, :integer
```
Comment (shanear) [inline db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb]: the `source_issue` is the original issue that an issue is appealing/reviewing

File: spec/feature/intake_spec.rb [hunk 12]
```diff
@@ -708,7 +708,7 @@
           expect(page).to have_content("Ineligible for Higher-Level Review")
         end
 
-        scenario "Start a RAMP refiling" do
+        scenario "Complete a RAMP refiling for an appeal" do
```
Comment (shanear) [inline spec/feature/intake_spec.rb]: Split out the feature tests into Appeal and non-appeal. This will get more interesting with future stories.

**Comment by nickheiner-usds** [comment]
image

> Established EP: undefined

Is that intended?

**Comment by nickheiner-usds** [comment]
Should it be possible to select all these boxes at once?

image

**Comment by nickheiner-usds** [comment]
> All the AC should be satisfied except:
> - If the selection includes at least one eligible contention, navigates to success confirmation page (#4064)

Are you sure this AC is not satisfied? It seems like it works to me. :Smile:

**Comment by nickheiner-usds** [comment]
> All selection requests (SC, HLR, appeal) from previous step navigate to this page.

I don't understand what this AC from the ticket means.

**Comment by nickheiner-usds** [comment]
> Verify that the list of contentions matches all contentions previously associated with that Veteran's RAMP EP (682 or 683).

Can we meaningfully test this in local dev mode?

File: app/models/ramp_refiling_intake.rb (hunk 4 above)
Comment (nickheiner-usds) [reply app/models/ramp_refiling_intake.rb]: :+1: A failure mode of "the transaction failed so you have to try again in a few minutes" is preferable to "data is corrupted".

File: app/models/ramp_review.rb (hunk 5 above)
Comment (nickheiner-usds) [reply app/models/ramp_review.rb]: Do we not want to share it?

File: client/app/intake/actions/rampRefiling.js [hunk 13]
```diff
@@ -97,3 +97,75 @@ export const setOutsideCaseflowStepsConfirmed = (isConfirmed) => ({
     }
   }
 });
+
+export const processFinishError = () => ({ type: ACTIONS.PROCESS_FINISH_ERROR });
+
+const validateSelectedIssues = (rampRefiling) => (
+  rampRefiling.hasIneligibleIssue || _.some(rampRefiling.issues, (issue) => issue.isSelected)
```
Comment (nickheiner-usds) [inline client/app/intake/actions/rampRefiling.js]:
```js
_.some(rampRefiling.issues, (issue) => issue.isSelected)
```

this can just be:

```js
_.some(rampRefiling.issues, (issue) => 'isSelected')
```

File: client/app/intake/actions/rampRefiling.js [hunk 14]
```diff
@@ -97,3 +97,75 @@ export const setOutsideCaseflowStepsConfirmed = (isConfirmed) => ({
     }
   }
 });
+
+export const processFinishError = () => ({ type: ACTIONS.PROCESS_FINISH_ERROR });
+
+const validateSelectedIssues = (rampRefiling) => (
```
Comment (nickheiner-usds) [inline client/app/intake/actions/rampRefiling.js]: The paren around the method body is unnecessary here.

File: client/app/intake/pages/rampRefiling/finish.jsx (hunk 9 above)
Comment (nickheiner-usds) [inline client/app/intake/pages/rampRefiling/finish.jsx]: Don't use manual boolean logic for `className`. Use `classNames` instead:

```js
classNames({
    'usa-input-error': issuesSelectedError
})
```

https://www.npmjs.com/package/classnames

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 15]
```diff
@@ -148,15 +167,57 @@ export default connect(
...
+    setOutsideCaseflowStepsConfirmed,
+    processFinishError
   }, dispatch)
 )(Finish);
 
+class FinishNextButton extends React.PureComponent {
+  handleClick = () => {
+    this.props.completeIntake(this.props.intakeId, this.props.rampRefiling).then(
+      (completeWasSuccessful) => {
+        if (completeWasSuccessful) {
```
Comment (nickheiner-usds) [inline client/app/intake/pages/rampRefiling/finish.jsx]: What do we do if the complete failed? Can we at least add a TODO or ticket to handle the error case? (Probably better not to bloat this PR.)

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 16]
```diff
@@ -148,15 +167,57 @@ export default connect(
...
+          this.props.history.push('/completed');
+        }
+      }
+    );
+  }
+
+  render = () =>
+    <Button
+      id="finish-intake"
+      name="finish-intake"
```
Comment (nickheiner-usds) [inline client/app/intake/pages/rampRefiling/finish.jsx]: Why is this necessary?

File: client/app/intake/pages/rampRefiling/finish.jsx [hunk 17]
```diff
@@ -148,15 +167,57 @@ export default connect(
...
+    rampRefiling
+  }),
+  (dispatch) => bindActionCreators({
+    completeIntake
+  }, dispatch)
+)(FinishNextButton);
+
 export class FinishButtons extends React.PureComponent {
-  render = () => <CancelButton />
+  render = () =>
```
Comment (nickheiner-usds) [inline client/app/intake/pages/rampRefiling/finish.jsx]: Once I land https://github.com/department-of-veterans-affairs/caseflow/pull/4276, you'll be able to just return an array of elements here instead of using a wrapper `div`. :smile:

File: client/app/intake/reducers/rampRefiling.js [hunk 18]
```diff
@@ -145,14 +165,68 @@ export const rampRefilingReducer = (state = mapDataToInitialRampRefiling(), acti
...
         $set: action.payload.isConfirmed
+      },
+      outsideCaseflowStepsError: {
+        $set: null
+      }
+    });
+  case ACTIONS.COMPLETE_INTAKE_STEPS_NOT_CONFIRMED:
+    return update(state, {
+      outsideCaseflowStepsError: {
+        $set: "You must confirm you've completed the steps"
```
Comment (nickheiner-usds) [inline client/app/intake/reducers/rampRefiling.js]: Do we want punctuation at the end of this message and the `NO_ISSUES_SELECTED_ERROR` message? I don't see copy listed in https://github.com/department-of-veterans-affairs/caseflow/issues/3966.

File: db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb (hunk 11 above)
Comment (nickheiner-usds) [reply db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb]: What table does the `source_issue_id` look up in? Is it `ramp_issues`?

File: client/app/intake/pages/rampRefiling/finish.jsx (hunk 8 above)
Comment (nickheiner-usds) [reply client/app/intake/pages/rampRefiling/finish.jsx]: `ref` is correct, but it is better not to define functions in `render` to save wasted renders. Here's the fix:

```diff
diff --git a/client/app/intake/pages/rampRefiling/finish.jsx b/client/app/intake/pages/rampRefiling/finish.jsx
index 1e5ac5033..c486bf52e 100644
--- a/client/app/intake/pages/rampRefiling/finish.jsx
+++ b/client/app/intake/pages/rampRefiling/finish.jsx
@@ -24,6 +24,8 @@ class Finish extends React.PureComponent {
     }
   }
 
+  getOutsideCaseflowStepsNode = (node) => this.outsideCaseflowStepsNode = node
+
   render() {
     const {
       rampRefilingStatus,
@@ -86,7 +88,7 @@ class Finish extends React.PureComponent {
     return <div>
       <h1>Finish processing RAMP Selection form</h1>
 
-      <ol className="cf-bare-list" ref={(node) => this.outsideCaseflowStepsNode = node}>
+      <ol className="cf-bare-list" ref={this.getOutsideCaseflowStepsNode}>
         <li>
           <div className="cf-intake-step">
             <strong>1. Complete the following tasks outside Caseflow and mark when complete</strong>
```

File: client/app/intake/actions/rampRefiling.js [hunk 19]
```diff
@@ -97,3 +97,75 @@ export const setOutsideCaseflowStepsConfirmed = (isConfirmed) => ({
...
+  }
+
+  dispatch({
+    type: ACTIONS.COMPLETE_INTAKE_START,
+    meta: { analytics }
+  });
+
+  const data = {
+    has_ineligible_issue: rampRefiling.hasIneligibleIssue,
+    issue_ids: _.reduce(rampRefiling.issues, (selectedIssues, issue) => {
```
Comment (nickheiner-usds) [inline client/app/intake/actions/rampRefiling.js]: `reduce` is low level, and it is rare to need to do it directly. In this case, we're just doing a `filter` and a `map`, so we can express that at a higher level:

```js
    issue_ids: _(rampRefiling.issues).
      filter('isSelected').
      map('id').
      value()
```

File: client/app/intake/actions/rampRefiling.js [hunk 20]
```diff
@@ -97,3 +97,75 @@ export const setOutsideCaseflowStepsConfirmed = (isConfirmed) => ({
...
 });
+
+export const processFinishError = () => ({ type: ACTIONS.PROCESS_FINISH_ERROR });
+
+const validateSelectedIssues = (rampRefiling) => (
+  rampRefiling.hasIneligibleIssue || _.some(rampRefiling.issues, (issue) => issue.isSelected)
+);
+
+export const completeIntake = (intakeId, rampRefiling) => (dispatch) => {
+  let hasError = false;
```
Comment (nickheiner-usds) [inline client/app/intake/actions/rampRefiling.js]: I think we can eliminate this variable to clearer, more direct code:

```diff
diff --git a/client/app/intake/actions/rampRefiling.js b/client/app/intake/actions/rampRefiling.js
index cf9310ce0..1c675ee65 100644
--- a/client/app/intake/actions/rampRefiling.js
+++ b/client/app/intake/actions/rampRefiling.js
@@ -105,15 +105,13 @@ const validateSelectedIssues = (rampRefiling) => (
 );
 
 export const completeIntake = (intakeId, rampRefiling) => (dispatch) => {
-  let hasError = false;
-
   if (!rampRefiling.outsideCaseflowStepsConfirmed) {
     dispatch({
       type: ACTIONS.COMPLETE_INTAKE_STEPS_NOT_CONFIRMED,
       meta: { analytics }
     });
 
-    hasError = true;
+    return Promise.resolve(false);
   }
 
   if (!validateSelectedIssues(rampRefiling)) {
@@ -122,10 +120,6 @@ export const completeIntake = (intakeId, rampRefiling) => (dispatch) => {
       meta: { analytics }
     });
 
-    hasError = true;
-  }
-
-  if (hasError) {
     return Promise.resolve(false);
   }
 

```

File: app/models/ramp_review.rb (hunk 5 above)
Comment (shanear) [reply app/models/ramp_review.rb]: Yep, and we do share it. That's why it's here in the abstract base class :)

**Comment by shanear** [comment]
> Should it be possible to select all these boxes at once and then hit submit?

Yes, they could have listed all 3 eligible contentions, and then wrote some bad contentions as well

**Comment by shanear** [comment]
> Are you sure this AC is not satisfied? It seems like it works to me. 😄

It navigates to a success page haha. But it's the RampElection success page, which is wrong. For now :)

File: client/app/intake/actions/rampRefiling.js (hunk 20 above)
Comment (shanear) [reply client/app/intake/actions/rampRefiling.js]: There are two error conditions being checked for in this action, which is why I added the variable. I agree the variable is a little janky, but this doesn't work.

File: client/app/intake/actions/rampRefiling.js (hunk 19 above)
Comment (shanear) [reply client/app/intake/actions/rampRefiling.js]: lol, I always reach for reduce. Yeah this is nicer. Will change

File: client/app/intake/pages/rampRefiling/finish.jsx (hunk 17 above)
Comment (shanear) [reply client/app/intake/pages/rampRefiling/finish.jsx]: yayy! I've always thought that was janky that you had to do that.

File: client/app/intake/pages/rampRefiling/finish.jsx (hunk 16 above)
Comment (shanear) [reply client/app/intake/pages/rampRefiling/finish.jsx]: it might not be lol. Ill try taking it out

File: db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb (hunk 11 above)
Comment (shanear) [reply db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb]: `ramp_issues`. You can see that here: https://github.com/department-of-veterans-affairs/caseflow/pull/4272/files#diff-eb0859e45ebdc926e4b14b99643b0f61R3

**Comment by shanear** [comment]
> Can we meaningfully test this in local dev mode?

We do by faking out contentions associated with the EP id on the ramp election, but we also need to test in UAT. I will work with alexis on that.

Since this is all behind a feature flag I don't feel the need to test it pre-merge

File: client/app/intake/pages/rampRefiling/finish.jsx (hunk 15 above)
Comment (shanear) [reply client/app/intake/pages/rampRefiling/finish.jsx]: https://github.com/department-of-veterans-affairs/caseflow/issues/4293

File: client/app/intake/reducers/rampRefiling.js (hunk 18 above)
Comment (shanear) [reply client/app/intake/reducers/rampRefiling.js]: The message in RampElection doesn't have a period, so i didn't put a period here.

I'll let @abbyraskinUSDS submit a polish issue to clean up these tweaks.

File: client/app/intake/actions/rampRefiling.js (hunk 20 above)
Comment (nickheiner-usds) [reply client/app/intake/actions/rampRefiling.js]: Ah, that's right. Ok, I missed that the dispatching would actually have a side-effect outside of this function, which was obvious. Good point.

File: client/app/intake/reducers/rampRefiling.js (hunk 18 above)
Comment (nickheiner-usds) [reply client/app/intake/reducers/rampRefiling.js]: :+1: for consistency.

File: app/models/ramp_review.rb (hunk 5 above)
Comment (nickheiner-usds) [reply app/models/ramp_review.rb]: :+1: When you said "word for word copied" I'd interpreted that to mean that it still existed in `ramp_election` as well, but I now see that's not the case.

File: app/models/ramp_review.rb [hunk 21]
```diff
@@ -33,8 +36,46 @@ def higher_level_review?
     HIGHER_LEVEL_REVIEW_OPTIONS.include?(option_selected)
   end
 
+  def create_end_product!
```
Comment (nickheiner-usds) [inline app/models/ramp_review.rb]: I'm sure you're aware that this method looks very similar to what you wrote in `establish_claim` in b42e4301. Do we want to try to share logic between the two?

File: db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb (hunk 11 above)
Comment (nickheiner-usds) [reply db/migrate/20180103204311_add_source_issue_id_to_ramp_issues.rb]: Ah, great, thanks.

File: spec/feature/intake_spec.rb [hunk 22]
```diff
@@ -774,14 +774,97 @@
...
+          expect(page).to have_content("Finish processing RAMP Selection form")
+          expect(page).to have_content("You must confirm you've completed the steps")
+
+          click_label("confirm-outside-caseflow-steps")
+
+          safe_click "#finish-intake"
+
+          expect(page).to have_content("Intake completed")
+
+          expect(Fakes::VBMSService).to_not have_received(:establish_claim!)
```
Comment (nickheiner-usds) [inline spec/feature/intake_spec.rb]: Can you remind me why we want this assertion?

File: spec/feature/intake_spec.rb [hunk 23]
```diff
@@ -774,14 +774,97 @@
...
+          expect(ramp_refiling.issues.count).to eq(2)
+          expect(ramp_refiling.issues.first.description).to eq("Left knee rating increase")
+          expect(ramp_refiling.issues.last.description).to eq("Left shoulder service connection")
+        end
+
+        scenario "Complete a RAMP Refiling for a supplemental claim" do
+          # Create an complete Higher level review RAMP election
```
Comment (nickheiner-usds) [inline spec/feature/intake_spec.rb]: nit: Let's either write it as "Higher Level Review" or "higher level review".

File: spec/models/ramp_issue_spec.rb [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+      it "sets description from source issue" do
+        subject
+        expect(ramp_issue.description).to eq("Source description")
+      end
+    end
+
+    context "description was set" do
+      let(:description) { "Fresh" }
+
+      it "sets description from source issue" do
```
Comment (nickheiner-usds) [inline spec/models/ramp_issue_spec.rb]: Don't we want this to say "**does not** set description from source issue"?

File: spec/models/ramp_refiling_spec.rb [hunk 25]
```diff
@@ -30,6 +30,28 @@
...
+
+    let(:source_issues) do
+      [
+        completed_ramp_election.issues.create!(description: "Firsties"),
+        completed_ramp_election.issues.create!(description: "Secondsies")
+      ]
+    end
+
+    let(:outdated_issue) do
+      ramp_refiling.issues.create!(description: "i will be destroyed")
```
Comment (nickheiner-usds) [inline spec/models/ramp_refiling_spec.rb]: uber-nit: can we make this an uppercase `I`?

File: spec/models/ramp_issue_spec.rb (hunk 24 above)
Comment (shanear) [reply spec/models/ramp_issue_spec.rb]: good catch 💯

File: spec/models/ramp_refiling_spec.rb (hunk 25 above)
Comment (shanear) [reply spec/models/ramp_refiling_spec.rb]: hahaha, I kind of like it lowercase, adds to the tragic nature of the test

File: spec/feature/intake_spec.rb (hunk 22 above)
Comment (shanear) [reply spec/feature/intake_spec.rb]: per our discussion, it's because Appeals don't create EPs

File: spec/models/ramp_refiling_intake_spec.rb [hunk 26]
```diff
@@ -121,6 +122,61 @@
     end
   end
 
+  context "#complete!" do
```
Comment (nickheiner-usds) [inline spec/models/ramp_refiling_intake_spec.rb]: :+1:
